ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Mempool/Init.hs

```diff
@@ -57,27 +56,32 @@ forkSyncStateOnTipPointChange ::
   ResourceRegistry m ->
   MempoolEnv m blk ->
   m ()
-forkSyncStateOnTipPointChange registry menv =
-  void $
+forkSyncStateOnTipPointChange topLevelRegistry menv = do
+  w <-
     forkLinkedWatcher
-      registry
+      (mpEnvRegistry menv)
       "Mempool.syncStateOnTipPointChange"
       Watcher
         { wFingerprint = id
         , wInitial = Nothing
         , wNotify = action
         , wReader = getCurrentTip
         }
+
+  -- With this allocation on the top level registry, we make sure that we first
+  -- stop the watcher thread before closing the mempool registry, as otherwise
+  -- we would run into a race condition (the thread might try to re-sync and
+  -- allocate a forker on the mempool registry which would be closing down).
+  void $ allocate topLevelRegistry (\_ -> pure w) cancelThread
```

Comment:
@jasagredo I came across this code while reviewing #1738. Is this code here still necessary with your recent work in https://github.com/IntersectMBO/io-classes-extra/pull/10?